File: frontend/src/api/projectCosts.ts
/**
 * Project Costs API — externals and OoP equipment per project/period.
 */
import { apiClient } from './client';

export interface ExternalResource {
  id: string;
  display_name: string;
  initials: string | null;
  employee_id: string;
}

export interface ExternalLine {
  id: string;
  tenant_id: string;
  project_id: string;
  period_id: string;
  resource_id: string | null;
  resource_name: string | null;
  description: string | null; // name / description (shown as OoP Resource when no resource_name)
  notes: string | null;        // additional free-text notes
  hours: number;
  rate: number;       // cents per hour
  total_cost: number; // cents
  created_by: string;
  created_at: string;
  updated_at: string;
  project_name?: string | null;
}

export interface ExternalLineCreate {
  project_id: string;
  period_id: string;
  resource_id?: string;
  description?: string; // name / description when no resource_id
  notes?: string;        // additional free-text notes
  hours: number;
  rate: number; // cents per hour
}

export interface ExternalLineUpdate {
  resource_id?: string;
  description?: string;
  notes?: string;
  hours?: number;
  rate?: number;
}

export interface EquipmentLine {
  id: string;
  tenant_id: string;
  project_id: string;
  period_id: string;
  description: string | null;
  cost: number; // cents
  created_by: string;
  created_at: string;
  updated_at: string;
  project_name?: string | null;
}

export interface EquipmentLineCreate {
  project_id: string;
  period_id: string;
  description?: string;
  cost: number; // cents
}

export interface EquipmentLineUpdate {
  description?: string;
  cost?: number;
}

export interface ProjectCostSummaryItem {
  project_id: string;
  project_name: string;
  externals_total: number;
  equipment_total: number;
  combined_total: number;
}

export interface ProjectCostSummary {
  externals_total: number;
  equipment_total: number;
  combined_total: number;
  per_project: ProjectCostSummaryItem[];
}

function buildQuery(params: Record<string, string | undefined>): string {
  const entries = Object.entries(params).filter(([, v]) => v !== undefined && v !== '');
  if (entries.length === 0) return '';
  return '?' + entries.map(([k, v]) => `${k}=${encodeURIComponent(v!)}`).join('&');
}

export const projectCostsApi = {
  async listExternalResources(): Promise<ExternalResource[]> {
    return apiClient.get<ExternalResource[]>('/project-costs/external-resources');
  },

  async listExternals(params: { period_id?: string; project_id?: string } = {}): Promise<ExternalLine[]> {
    return apiClient.get<ExternalLine[]>(`/project-costs/externals${buildQuery(params)}`);
  },

  async createExternal(data: ExternalLineCreate): Promise<ExternalLine> {
    return apiClient.post<ExternalLine>('/project-costs/externals', data);
  },

  async updateExternal(id: string, data: ExternalLineUpdate): Promise<ExternalLine> {
    return apiClient.put<ExternalLine>(`/project-costs/externals/${id}`, data);
  },

  async deleteExternal(id: string): Promise<void> {
    await apiClient.delete(`/project-costs/externals/${id}`);
  },

  async listEquipment(params: { period_id?: string; project_id?: string } = {}): Promise<EquipmentLine[]> {
    return apiClient.get<EquipmentLine[]>(`/project-costs/equipment${buildQuery(params)}`);
  },

  async createEquipment(data: EquipmentLineCreate): Promise<EquipmentLine> {
    return apiClient.post<EquipmentLine>('/project-costs/equipment', data);
  },

  async updateEquipment(id: string, data: EquipmentLineUpdate): Promise<EquipmentLine> {
    return apiClient.put<EquipmentLine>(`/project-costs/equipment/${id}`, data);
  },

  async deleteEquipment(id: string): Promise<void> {
    await apiClient.delete(`/project-costs/equipment/${id}`);
  },

  async getSummary(params: { period_id?: string; project_id?: string } = {}): Promise<ProjectCostSummary> {
    return apiClient.get<ProjectCostSummary>(`/project-costs/summary${buildQuery(params)}`);
  },
};

File: api/app/models/project_costs.py
"""Project cost models - externals and OoP equipment per project/period."""
import uuid
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Integer, DateTime, Text, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column

from api.app.db.base import Base


def generate_uuid() -> str:
    return str(uuid.uuid4())


class ProjectExternalLine(Base):
    """External contractor/vendor cost entered by PM per project and period."""
    __tablename__ = "project_external_lines"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    tenant_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), nullable=False)
    period_id: Mapped[str] = mapped_column(String(36), ForeignKey("periods.id"), nullable=False)

    resource_id: Mapped[Optional[str]] = mapped_column(String(36), ForeignKey("resources.id"), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # name / description
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)        # additional notes
    cost: Mapped[int] = mapped_column(Integer, nullable=False)  # cents, flat cost

    created_by: Mapped[str] = mapped_column(String(36), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    __table_args__ = (
        Index("ix_ext_tenant_period", "tenant_id", "period_id"),
    )


class ProjectEquipmentLine(Base):
    """OoP equipment cost entered by PM per project and period."""
    __tablename__ = "project_equipment_lines"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    tenant_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), nullable=False)
    period_id: Mapped[str] = mapped_column(String(36), ForeignKey("periods.id"), nullable=False)

    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    cost: Mapped[int] = mapped_column(Integer, nullable=False)  # cents

    created_by: Mapped[str] = mapped_column(String(36), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    __table_args__ = (
        Index("ix_equip_tenant_period", "tenant_id", "period_id"),
    )
